AST: Don't optimize decls that are unavailable in extensions

tshortli opened the pull request:
Unavailable decl optimization is meant to optimize declarations that are unreachable at runtime. A declaration that is marked unavailable in app extensions (e.g. `@available(macOSApplicationExtension, unavailable)`) may be reachable by non-extension processes so it cannot be safely optimized.

Resolves rdar://122924862

tshortli committed:
AST: Don't optimize decls that are unavailable in extensions.

Unavailable decl optimization is meant to optimize declarations that are
unreachable at runtime. A declaration that is marked unavailable in app
extensions (e.g. `@available(macOSApplicationExtension, unavailable)`) may be
reachable by non-extension processes so it cannot be safely optimized.

Resolves rdar://122924862

tshortli commented:
@swift-ci please smoke test

beccadax commented on lib/AST/Availability.cpp:
```diff
@@ -321,11 +321,29 @@ llvm::Optional<AvailableAttrDeclPair> Decl::getSemanticUnavailableAttr() const {
+  // If the decl is only unavailable to app extensions it still may need to be
+  // present at runtime for non-extension processes.
+  if (isApplicationExtensionPlatform(unavailableAttr->Platform))
+    return false;
```
When this happens, should we look for a non-app extension availability attribute instead of just returning `false`? Think of a test case like:

```swift
@available(macOS, unavailable)
@available(macOSApplicationExtension, unavailable)
public func doublyUnavailableOnMacOSFunc() {}
```

(If you want to do that, a better approach might be to make `getSemanticUnavailableAttr()` take a flag to ignore app extension platforms.)

tshortli replied:
I considered that, but I really think we shouldn't accept the code in your example. Having both attributes is redundant, and rather than complicating querying I think the AST should be guaranteed to not have this situation.

beccadax replied:
*Do* we reject the code, though? And are there no situations—for instance, when we inherit availability attributes from another platform—where this can happen in a more reasonable example?

```
// Which one of these do we even use if you're building for Mac Catalyst with -app-extension???
@available(macCatalyst, unavailable)
@available(iOSApplicationExtension, unavailable)
public func confusingDiamondAvailabilityInheritanceFunc() {}
```

tshortli replied:
Ok, it took a fair bit of plumbing of new parameters but I've handled this case.

tshortli committed:
AST: Ignore availability attrs for app extension platforms in some queries.

When determining whether a declaration should be considered unavailable at
runtime, ignore `@available` attributes for application extension platforms but
continue searching for other `@available` attributes that might still make the
declaration unavailable. This ensures corner cases like these are handled:

```
// Dubious, but allowed
@available(macOS, unavailable)
@available(macOSApplicationExtension, unavailable)
public func doublyUnavailableOnMacOSFunc() {}

// Expresses an uncommon, but valid constraint
@available(macCatalyst, unavailable)
@available(iOSApplicationExtension, unavailable)
public func confusingDiamondAvailabilityInheritanceFunc() {}
```

tshortli commented:
@swift-ci please test
